### PR TITLE
Write 'Best practices'

### PR DESCRIPTION
**Pull request by mstephen19**
Migrated [this](https://gitlab.com/apify-public/wiki/-/blob/master/wiki/Writing-actors/How-to-write-and-not-write-an-actor.md)

### CODE REVIEW

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+Don't write code lke this:
+
+```JavaScript
+const x = (y) => (y - 32) * (5 / 9);
+```
+
+That is quite confusing due to the undescriptive naming and the magic numbers. Do this instead:
+
+```JavaScript
+// Converts a fahrenheit value to celcius
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:42]:
```suggestion
// Converts a fahrenheit value to celsius
```

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+const fahrenheitToCelcius = (celcius) => (celcius - 32) * (5 / 9);
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:43]:
```suggestion
const fahrenheitToCelsius = (celsius) => (celsius - 32) * (5 / 9);
```

**Comment by mstephen19** [comment]
View preview [here](https://developers.apify.com/academy/web-scraping-for-beginners/best-practices?version=concepts/best-practices)

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+### [](#clean-code) Clean code
+
+Praise [clean code](https://blog.risingstack.com/javascript-clean-coding-best-practices-node-js-at-scale/)! Use proper variable and function names that are descriptive of what they are, and split your code into smaller [pure](https://en.wikipedia.org/wiki/Pure_function) functions.
+
+### [](#constants) Constant variables
+
+Define any [constant variables](https://softwareengineering.stackexchange.com/questions/250619/best-practices-reasons-for-string-constants-in-javascript) that globally apply to the scraper in a single file named **constants.js**, from where they will all be imported. Constant variable names should be in `UPPERCASE_WITH_UNDERSCORES` style.
+
+> If you have a whole lot of constant variables, they can be in a folder named **constant** organized into different files.
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:27]:
```suggestion
> If you have a whole lot of constant variables, they can be in a folder named **constants** organized into different files.
```

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+If you're writing your scraper in JavaScript, use [ES6](https://www.w3schools.com/js/js_es6.asp) features and ditch the old ones which they replace. This means using `const` and `let` instead of `var`, `includes` instead of `indexOf`, etc.
+
+> To learn more about some of the most popular (and awesome) ES6+ features, check out [this](https://medium.com/@matthiasvstephens/why-is-es6-so-awesome-88bff6857849) article.
+
+### [](#no-magic-numbers) No magic numbers
+
+Avoid using [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) as much as possible. Either declare them as a **constant** variable in your **constants.js** file, or if they are only used once, add a comment explaining what the number is.
+
+Don't write code lke this:
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:39]:
```suggestion
Don't write code like this:
```

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+That is quite confusing due to the undescriptive naming and the magic numbers. Do this instead:
+
+```JavaScript
+// Converts a fahrenheit value to celsius
+const fahrenheitToCelsius = (celsius) => (celsius - 32) * (5 / 9);
+```
+
+> Additionally, try to make values for options such as `maxRequestRetries` in the Apify SDK configurable via input. If they absolutely must be a strict value, make them a constant variable.
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:52]: I think this is a bit out of place here in the beginners tutorial. I would just get rid of it, I don't think it's super important.

File: content/academy/web_scraping_for_beginners/best_practices.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+There are two main best practices when it comes to accepting input into a scraper.
+
+### [](#set-limits) Set limits
+
+When allowing your users to pass input properties which could break the scraper (such as **timeout** set to **0**), be sure to disallow ridiculous values. Set a maximum/minimum number allowed, maximum array input length, etc.
+
+### [](#validate) Validate
+
+Validate the input provided by the user! This should be the very first thing your scraper does. If the fields in the input are missing or in an incorrect type/format, either parse the value and correct it programmatically or throw an informative error telling the user how to fix the error.
```
Comment (mnmkng) [inline content/academy/web_scraping_for_beginners/best_practices.md:90]: I would add a callout that says that on the apify platform you can use input schema (with a link) to validate inputs and also generate a nice UI

**Review by mnmkng — CHANGES_REQUESTED** [review]

**Comment by metalwarrior665** [comment]
Haha, that's an old one but I guess most of it is non-controversial :)

File: content/academy/web_scraping_for_beginners/best_practices.md (hunk 5 above)
Comment (metalwarrior665) [reply content/academy/web_scraping_for_beginners/best_practices.md:52]: I would also skip it, I think right now the best practice is to not put as much stuff into input.

**Review by metalwarrior665 — APPROVED** [review]

**Review by mnmkng — APPROVED** [review]

**Pull request merged**